test/testdata/desugar/block_given.rb

```diff
@@ -0,0 +1,54 @@
+# typed: false
+
+class SomethingElse; end
+
+def method_with_block_param(&the_block_name)
+  # Valid cases that we should detect as special calls to `block_given?`
+  block_given?
+  self.block_given?
+  Kernel.block_given?
+  ::Kernel.block_given?
+
+  # Due to the current structure of Desugar.cc, we still detect this as a special `block_given?` call.
+  block_given? { "with a block?!" }
+  #            ^^^^^^^^^^^^^^^^^^^^ error: No body in block
+
+  # Cases we should not detect as special calls to `block_given?`
+  Object.block_given? # Normally not allowed: private method `block_given?' called for Object:Class
+  SomethingElse.block_given?
+  block_given?("with parameter")
+end
+
+def method_with_anonymous_block_param(&)
+  # Valid cases that we should detect as special calls to `block_given?`
+  block_given?
+  self.block_given?
+  Kernel.block_given?
+  ::Kernel.block_given?
+
+  # Due to the current structure of Desugar.cc, we still detect this as a special `block_given?` call.
+  block_given? { "with a block?!" }
+  #            ^^^^^^^^^^^^^^^^^^^^ error: No body in block
+
+  # Cases we should not detect as special calls to `block_given?`
+  Object.block_given? # Normally not allowed: private method `block_given?' called for Object:Class
+  SomethingElse.block_given?
+  block_given?("with parameter")
+end
+
+def method_with_implicit_block_param()
```

Comment:
It looks like this `dctx.enclosingBlockParamName.exists()` check was dead code all along:

https://github.com/sorbet/sorbet/blob/0fbadd4c5550e9bc76db624de755c58260d89a8d/ast/desugar/Desugar.cc#L946

By this point in the desugarer, any method that didn't have an explicit block argument, would have had a synthetic `&<blk>` added.

https://github.com/sorbet/sorbet/blob/0fbadd4c5550e9bc76db624de755c58260d89a8d/ast/desugar/Desugar.cc#L375-L382

Reply:
https://github.com/sorbet/sorbet/pull/9726